hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -0,0 +1,338 @@
+        for (Entry<Byte, Map<Byte, Long>> outerEntry : postponeTimeMap.entrySet()) {
+            for (Byte innerKey : outerEntry.getValue().keySet()) {
+                int innerOrdinal = leftTimeExtractors.containsKey(innerKey) ? 0 : 1;
+                int outerOrdinal = leftTimeExtractors.containsKey(outerEntry.getKey()) ? 0 : 1;
+                // innerOrdinal == outerOrdinal if the time bound is between timestamps on the same input, we ignore those
+                if (innerOrdinal != outerOrdinal) {
+                    found[innerOrdinal] = true;
+                }
+            }
+        }
+        if (!found[0] || !found[1]) {
+            throw new IllegalArgumentException("Not enough time bounds in postponeTimeMap");
+        }
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) throws Exception {
+        this.evalContext = ExpressionEvalContext.from(context);
+        SerializationService ss = evalContext.getSerializationService();
+        emptyLeftRow = new JetSqlRow(ss, new Object[columnCounts.f0()]);
+        emptyRightRow = new JetSqlRow(ss, new Object[columnCounts.f1()]);
+    }
+
+    @Override
+    public boolean tryProcess(int ordinal, @Nonnull Object item) {
+        assert ordinal == 0 || ordinal == 1; // bad DAG
+        if (!pendingOutput.isEmpty()) {
+            return processPendingOutput();
+        }
+
+        // drop the event, if it's late
+        for (Entry<Byte, ToLongFunctionEx<JetSqlRow>> en : timeExtractors(ordinal).entrySet()) {
+            long wmValue = lastReceivedWm.getValue(en.getKey());
+            long time = en.getValue().applyAsLong((JetSqlRow) item);
+            if (time < wmValue) {
+                logLateEvent(getLogger(), wmValue, item);
+                return true;
+            }
+        }
+
+        // if the item is not late, but would already be removed from the buffer, don't add it to the buffer
+        for (Entry<Byte, ToLongFunctionEx<JetSqlRow>> en : timeExtractors(ordinal).entrySet()) {
+            long joinTimeLimit = wmState.get(en.getKey());
+            long time = en.getValue().applyAsLong((JetSqlRow) item);
+            if (time < joinTimeLimit) {
+                if (!joinInfo.isInner()) {
+                    JetSqlRow joinedRow = composeRowWithNulls((JetSqlRow) item, ordinal);
+                    if (joinedRow != null && !tryEmit(joinedRow)) {
+                        pendingOutput.add(joinedRow);
+                        return false;
+                    }
+                }
+                return true;
+            }
+        }
+
+        // having side input, traverse the opposite buffer
+        if (currItem == null) {
+            currItem = (JetSqlRow) item;
+            buffer[ordinal].add(currItem);
+            iterator = buffer[1 - ordinal].iterator();
+            if (!joinInfo.isInner()) {
+                unusedEventsTracker[1 - ordinal].add(currItem);
+                unusedEventsTracker[ordinal].add(currItem);
```

Comment:
This is even worse, if we're left-joining, we should track only unused rows from the left.